nixos/xserver: Changed xrandrHeads to support corresponding monitor section configuration in Xorg

CMCDragonkai opened the pull request:
See previous PR: https://github.com/NixOS/nixpkgs/pull/13916

> This is a breaking change (actually it is no longer breaking) to `services.xserver.xrandrHeads`, however it is a necessary change. This allows xrandrHeads to support per-monitor section configuation in Xorg.conf. There's currently no other way to set the primary monitor or per monitor section configuration when using the `xrandrHeads` option.
> 
> Requested by: https://github.com/NixOS/nixpkgs/issues/7463
> 
> What it fixes:
> 
> When you multiple monitors setup. By default, the first monitor detected by `xrandr` is set to the primary monitor. However this may not be the monitor you need to be set as primary. In fact this monitor set to primary may in fact be disconnected. Which is exactly what happened to me. This has affected 3 programs for me:
> 1. XMonad  - gets confused with `Super + {w,e,r}`
> 2. SDDM - puts the login screen on the wrong monitor, and does not currently duplicate the login screen on all monitors
> 3. XMobar - puts the xmobar on the wrong monitor, as it only puts the taskbar on the primary monitor
> 
> This PR fixes all of the above 3 problems (Xmonad is no longer confused, SDDM puts the login screen on my primary monitor (laptop), Xmobar is working. It is simple to use. And simple to change. Previously you would do something like:
> 
> ```
> services.xserver.xrandrHeads = [ "DP-0" "HDMI-0" ];
> ```
> 
> Now you can do:
> 
> ```
> services.xserver.xrandrHeads = {
>     DP-0 =  ''
>         Option "Primary" "true"
>     '';
>     HDMI-0 = "";
> };
> ```
> 
> The reason why I did this, is to allow other per-monitor configuration options when using xrandrHeads (such as per-monitor DPI in the future). Note that the `services.xserver.monitorSection` setting does not work in conjunction with `xrandrHeads`.
> 
> I have tested this in current 15.09-small and it works.

Current usage is actually compatible with existing configuration. Here's the example:

```
services.xserver.xrandrHeads = [ "HDMI-0" { output = "DVI-0"; primary = true; monitorConfig = ""; } ];
```

CMCDragonkai committed:
nixos/xserver: Changed xrandrHeads to support corresponding monitor section configuration in Xorg

CMCDragonkai commented:
Waiting on https://github.com/NixOS/nixpkgs/pull/14053 to see what changes I need to make to this PR.

vcunat commented:
So this PR is still waiting, right?

CMCDragonkai commented:
@vcunat Yep, the thing is that @aszlig wants to have proper submodule support that is nested inside `either`. Currently this is not working, and options and documentation don't show up. So once that's done, this PR's implementation utilise it.

CMCDragonkai commented:
This PR just broke. @aszlig did you change something with the submodule type?

CMCDragonkai commented:
The PR that this PR is waiting on is taking too long. And also this change is forwards compatible when that PR does arrive. Can we just merge this? @vcunat 

Since no one has mentioned this, I want to point this out:

> By the way, there doesn't seem to be a way to making monitorConfig optional. Even with mkOption and default setting.

@domenkozar Also what changelog is required here?

jeremejevs commented:
I think comment edits don't trigger mention alerts.

@domenkozar Can you take a look look at @CMCDragonkai's question, please?

CMCDragonkai commented:
@domenkozar I have added an item into the 17.09 change log for this PR. Please check. Also how do you write code markup inside the change log XML?

domenkozar commented:
>  Also how do you write code markup inside the change log XML?

Example:

```
<programlisting><![CDATA[
{
  imports = [ <nixpkgs/nixos/modules/services/misc/gitit.nix> ];
}
]]></programlisting>
```

CMCDragonkai commented:
@domenkozar Ok, I don't actually need to write any code inside the XML. So is the PR mergeable?

jeremejevs commented on nixos/doc/manual/release-notes/rl-1709.xml:
```diff
@@ -17,7 +17,14 @@ has the following highlights: </para>
     </para>
   </listitem>
-
+  <listitem>
```
I believe this needs `<para></para>`?

CMCDragonkai replied:
Good catch!

basvandijk commented on nixos/modules/services/x11/xserver.nix:
```diff
@@ -33,12 +33,30 @@ let
+                monitorConfig = "";
+              }
+            );
+          primariness = if alreadyPrimary then true else processedHead.primary;
```
This could be simplified to:

```
primariness = alreadyPrimary || processedHead.primary;
```

aszlig commented on nixos/modules/services/x11/xserver.nix:
```diff
@@ -329,12 +351,27 @@ in
+          (submodule {
+            options.output = str;
+            options.primary = bool;
+            options.monitorConfig = lines;
```
These are not valid (sub)module options. As your branch is pushable, I'll try to fix that and get it in regardless of whether we have proper module support for this (as noted by https://github.com/NixOS/nixpkgs/pull/13916#issuecomment-198709244) because I currently don't have time to work on addressing #14053.

CMCDragonkai replied:
Ok what are the changes then?

CMCDragonkai replied:
I switched on Allow edits from maintainers so feel free to edit what you need.

aszlig committed:
nixos/xserver: Fix up/refactor xrandrHeads option

Using invalid module options in the submodule isn't very nice, because
it doesn't give very useful errors in case of type mismatch, also we
don't get descriptions of these options as they're effecively
nonexistent to the module system. Another downside of this is that
merging of these options isn't done correctly as well (eg. for
types.lines).

So we now have proper submodules for each xrandrHead and we also use
corcedTo in the type of xrandrHeads so that we can populate the
submodule's "output" option in case a plain string is defined for a list
item.

Instead of silently skipping multiple primary heads, we now have an
assertion, which displays a message and aborts configuration evaluation
appropriately.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig commented:
Okay, finished refactoring.
@basvandijk, @jeremejevs and of course @CMCDragonkai: Can you please review this again?

CMCDragonkai commented:
LGTM but is there documentation on this new coercedTo and your usage of `apply`? The function passed into `coercedTo`, what is it exactly doing?

aszlig committed:
nixos/doc: Improve release notes for xrandrHeads

The xrandrHeads option has been there since a long time, so there is no
need to advertise it as a new feature.

Instead, let's focus on just what has changed, which is that we now
assign one head to be primary.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/xserver: Document xrandrHeads.apply

It was asked by @CMCDragonkai to elaborate on that, so let's just do
this by actually providing a code comment.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig commented:
@CMCDragonkai: Just added a comment explaining the apply function. `coercedTo` on the other hand seems to be undocumented right now (at least I can't find it, maybe @abbradar can?), but it basically works like this:

The first argument is the type that needs to be coerced into the type specified by the third argument, the second argument is the actual coercion function which takes the former type as its argument and returns the final type.

Here I'm using it to convert plain strings into the submodule for `xrandrOptions`.

jeremejevs commented:
Can somebody trigger a rebuild and merge this, please?

CMCDragonkai commented:
Those automated builds are pretty useless. Best way to verify bug-free 
is to cherry pick the PR and test it on your machine. I was planning to, 
but haven't got around to it.


On 2/05/2017 6:58 AM, Olegs Jeremejevs wrote:
>
> Can somebody trigger a rebuild and merge this, please?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/NixOS/nixpkgs/pull/15353#issuecomment-298430915>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AAnHHamm4nzc-rnHOv9gFrhPteCIqNVgks5r1kd3gaJpZM4IbAGf>.
>

-- 
https://github.com/CMCDragonkai
+61420925975

CMCDragonkai commented:
@aszlig thanks!
